### PR TITLE
Add support for finetune guard classifier

### PR DESCRIPTION
**Pull request by VibhuJawa**
This PR  adds support for FineTuneGuard Model

### CODE REVIEW

File: nemo_curator/classifiers/aegis.py [hunk 1]
```diff
@@ -297,3 +360,122 @@ def _run_classifier(self, dataset: DocumentDataset) -> DocumentDataset:
...
+    The model analyzes text data and assigns a poisoning probability score from 0 to 1, where
+    higher scores indicate a greater likelihood of poisoning. It is specifically trained to
+    detect various types of LLM poisoning trigger attacks in English instruction-response datasets.
+
+    Model Capabilities:
+    - Trained on multiple known poisoning attack patterns
+    - Demonstrated strong zero-shot detection capabilities on novel attacks
+    - Particularly effective at identifying trigger patterns in partially poisoned datasets
+
+    Dataset Format:
```
Comment (VibhuJawa) [inline nemo_curator/classifiers/aegis.py:382]: TODO: Emphasize more about english

File: nemo_curator/classifiers/aegis.py [hunk 2]
```diff
@@ -69,36 +73,99 @@ class AegisConfig:
...
+
 class AegisModel(nn.Module):
     def __init__(
         self,
         pretrained_model_name_or_path: str,
         peft_model_name_or_path: str,
         dtype: torch.dtype,
-        token: str,
+        token: Optional[Union[str, bool]],
+        add_fintune_gaurd: bool = False,
```
Comment (sarahyurick) [inline nemo_curator/classifiers/aegis.py:109]:
```suggestion
        add_finetune_guard: bool = False,
```

File: nemo_curator/classifiers/aegis.py [hunk 3]
```diff
@@ -69,36 +73,99 @@ class AegisConfig:
...
+        autocast: bool = False,
     ):
         super().__init__()
         base_model = AutoModelForCausalLM.from_pretrained(
             pretrained_model_name_or_path, torch_dtype=dtype, token=token
         )
         self.model = PeftModel.from_pretrained(base_model, peft_model_name_or_path)
+        self.autocast = autocast
+        self.add_fintune_gaurd = add_fintune_gaurd
```
Comment (sarahyurick) [inline nemo_curator/classifiers/aegis.py:118]:
```suggestion
        self.add_finetune_guard = add_finetune_guard
```

File: nemo_curator/classifiers/aegis.py [hunk 4]
```diff
@@ -69,36 +73,99 @@ class AegisConfig:
...
+        if self.add_fintune_gaurd:
```
Comment (sarahyurick) [inline nemo_curator/classifiers/aegis.py:119]:
```suggestion
        if self.add_finetune_guard:
```

File: nemo_curator/classifiers/aegis.py [hunk 5]
```diff
@@ -69,36 +73,99 @@ class AegisConfig:
...
 
     @torch.no_grad()
-    def forward(self, batch):
-        response = self.model.generate(
-            **batch,
-            max_new_tokens=100,
-            pad_token_id=0,
-        )
+    def _forward(self, batch):
+        if self.add_fintune_gaurd:
```
Comment (sarahyurick) [inline nemo_curator/classifiers/aegis.py:124]:
```suggestion
        if self.add_finetune_guard:
```

File: nemo_curator/classifiers/aegis.py [hunk 6]
```diff
@@ -69,36 +73,99 @@ class AegisConfig:
...
 class AegisHFModel(HFModel):
     def __init__(self, config: AegisConfig, max_mem_gb: Optional[int] = None):
         self.config = config
         if max_mem_gb is None:
             max_mem_gb = _get_suggest_memory_for_classifier()
 
+        if self.config.add_finetune_guard:
+            if self.config.finetune_guard_path is None:
+                raise ValueError(
+                    "finetune_guard_path must be provided if add_fine_guard is True"
```
Comment (sarahyurick) [inline nemo_curator/classifiers/aegis.py:166]:
```suggestion
                    "finetune_guard_path must be provided if add_finetune_guard is True"
```

File: nemo_curator/classifiers/aegis.py [hunk 7]
```diff
@@ -111,11 +178,18 @@ def __init__(self, config: AegisConfig, max_mem_gb: Optional[int] = None):
...
         model = AegisModel(
-            self.config.pretrained_model_name_or_path,
-            self.config.peft_model_name_or_path,
-            self.config.dtype,
-            self.config.token,
+            pretrained_model_name_or_path=self.config.pretrained_model_name_or_path,
+            peft_model_name_or_path=self.config.peft_model_name_or_path,
+            dtype=self.config.dtype,
+            token=self.config.token,
+            add_fintune_gaurd=self.config.add_finetune_guard,
```
Comment (sarahyurick) [inline nemo_curator/classifiers/aegis.py:185]:
```suggestion
            add_finetune_guard=self.config.add_finetune_guard,
```

**Review by sarahyurick — COMMENTED** [review]
Added a few typo fixes but looks good so far!

When you get a chance can you add it to the `examples/` and `nemo_curator/scripts/` folders? And to the documentation? You should be able to reference https://github.com/NVIDIA/NeMo-Curator/pull/361 which has all the files that should be created/updated.

File: nemo_curator/classifiers/aegis.py (hunk 1 above)
Comment (VibhuJawa) [reply nemo_curator/classifiers/aegis.py:400]: Done

**Comment by VibhuJawa** [comment]
> When you get a chance can you add it to the examples/ and nemo_curator/scripts/ folders? And to the documentation? You should be able to reference https://github.com/NVIDIA/NeMo-Curator/pull/361 which has all the files that should be created/updated.

We can do that as a followup, as that is documentation related. Lets get this in first.

File: nemo_curator/classifiers/aegis.py [hunk 8]
```diff
@@ -41,6 +43,8 @@ class AegisConfig:
     pretrained_model_name_or_path: str = "meta-llama/LlamaGuard-7b"
     dtype: torch.dtype = torch.bfloat16
     max_length: int = 4096
+    add_finetune_guard: bool = False
+    finetune_guard_path: Optional[str] = None
```
Comment (sarahyurick) [inline nemo_curator/classifiers/aegis.py:47]:
```suggestion
    finetune_guard_path: str = "nvidia/FineTune-Guard"
```

File: nemo_curator/classifiers/aegis.py [hunk 9]
```diff
@@ -69,36 +73,99 @@ class AegisConfig:
...
+    def __init__(self, input_dim, dropout=0.7):
+        super().__init__()
+        self.input_dim = input_dim
+        self.dropout = Dropout(dropout)
+        self.sigmoid = torch.nn.Sigmoid()
+        self.input_layer = Linear(input_dim, input_dim)
+
+        self.hidden_layer_0 = Linear(input_dim, 2000)  # input_dim, 2000
+        self.hidden_layer_1 = Linear(2000, 500)  # 2000, 100
+        self.hidden_layer_2 = Linear(500, 1)  # 1000, 100
```
Comment (sarahyurick) [inline nemo_curator/classifiers/aegis.py:86]:
```suggestion
        self.hidden_layer_0 = Linear(input_dim, 2000)
        self.hidden_layer_1 = Linear(2000, 500)
        self.hidden_layer_2 = Linear(500, 1)
```

File: nemo_curator/classifiers/aegis.py [hunk 10]
```diff
@@ -69,36 +73,99 @@ class AegisConfig:
...
     @torch.no_grad()
-    def forward(self, batch):
-        response = self.model.generate(
-            **batch,
-            max_new_tokens=100,
-            pad_token_id=0,
-        )
+    def _forward(self, batch):
+        if self.add_finetune_guard:
+            # output_hidden_states=True, return_dict_in_generate=True, max_new_tokens=0, pad_token_id=0)
```
Comment (sarahyurick) [inline nemo_curator/classifiers/aegis.py:125]:
```suggestion
```

File: nemo_curator/classifiers/aegis.py [hunk 11]
```diff
@@ -69,36 +73,99 @@ class AegisConfig:
...
+                return_dict_in_generate=True,
+            )
+            # Access the hidden state of the last non-generated token from the last layer
+            fine_guard_input_tensor = response.hidden_states[0][32][:, -1, :].to(
+                torch.float
+            )
+            fine_guard_output_tensor = self.finetune_guard_net(
+                fine_guard_input_tensor
+            ).flatten()
+            return fine_guard_output_tensor
```
Comment (sarahyurick) [inline nemo_curator/classifiers/aegis.py:140]:
```suggestion
            finetune_guard_input_tensor = response.hidden_states[0][32][:, -1, :].to(
                torch.float
            )
            finetune_guard_output_tensor = self.finetune_guard_net(
                finetune_guard_input_tensor
            ).flatten()
            return finetune_guard_output_tensor
```

File: nemo_curator/classifiers/aegis.py [hunk 12]
```diff
@@ -69,36 +73,99 @@ class AegisConfig:
...
         self.config = config
         if max_mem_gb is None:
             max_mem_gb = _get_suggest_memory_for_classifier()
 
+        if self.config.add_finetune_guard:
+            if self.config.finetune_guard_path is None:
+                raise ValueError(
+                    "finetune_guard_path must be provided if add_finetune_guard is True"
+                )
+
```
Comment (sarahyurick) [inline nemo_curator/classifiers/aegis.py:168]:
```suggestion
```

File: nemo_curator/classifiers/aegis.py [hunk 13]
```diff
@@ -297,3 +375,129 @@ def _run_classifier(self, dataset: DocumentDataset) -> DocumentDataset:
...
+    Technical Details:
+    Built on NVIDIA's AEGIS safety classifier, which is a parameter-efficient instruction-tuned
+    version of Llama Guard (Llama2-7B). Access to the base Llama Guard model on HuggingFace
+    (https://huggingface.co/meta-llama/LlamaGuard-7b) is required via a user access token.
+    """
+
+    def __init__(
+        self,
+        finetune_guard_path: Optional[str] = None,
```
Comment (sarahyurick) [inline nemo_curator/classifiers/aegis.py:421]:
```suggestion
```

File: nemo_curator/classifiers/aegis.py [hunk 14]
```diff
@@ -297,3 +375,129 @@ def _run_classifier(self, dataset: DocumentDataset) -> DocumentDataset:
...
+        max_chars: int = 6000,
+        autocast: bool = True,
+        device_type: str = "cuda",
+        max_mem_gb: Optional[int] = None,
+    ):
+        """
+        Constructs the classifier
+
+        Args:
+            finetune_guard_path (Optional[str]): The path to the fine-tune guard model
```
Comment (sarahyurick) [inline nemo_curator/classifiers/aegis.py:436]:
```suggestion
```

File: nemo_curator/classifiers/aegis.py [hunk 15]
```diff
@@ -297,3 +375,129 @@ def _run_classifier(self, dataset: DocumentDataset) -> DocumentDataset:
...
+                                it defaults to the available GPU memory minus 4 GB.
+
+        """
+
+        _aegis_variant = "nvidia/Aegis-AI-Content-Safety-LlamaGuard-Defensive-1.0"
+        config = AegisConfig(
+            peft_model_name_or_path=_aegis_variant,
+            token=token,
+            add_finetune_guard=True,
+            finetune_guard_path=finetune_guard_path,
```
Comment (sarahyurick) [inline nemo_curator/classifiers/aegis.py:460]:
```suggestion
```

File: nemo_curator/classifiers/aegis.py [hunk 16]
```diff
@@ -297,3 +375,129 @@ def _run_classifier(self, dataset: DocumentDataset) -> DocumentDataset:
...
+            batch_size=batch_size,
+            out_dim=1,
+            pred_column=self._prob_column,
+            max_chars=max_chars,
+            device_type=device_type,
+            autocast=autocast,
+        )
+
+    def _run_classifier(self, dataset: DocumentDataset):
+        print("Starting AEGIS classifier inference", flush=True)
```
Comment (sarahyurick) [inline nemo_curator/classifiers/aegis.py:488]:
```suggestion
        print("Starting FineTune-Guard classifier inference", flush=True)
```

**Review by sarahyurick — CHANGES_REQUESTED** [review]

**Review by sarahyurick — APPROVED** [review]

**Comment by sarahyurick** [comment]
Some of these commits are not signed, so re-opening it as https://github.com/NVIDIA/NeMo-Curator/pull/397.

**Pull request closed without merging**